### ISSUE TITLE
Error when reading json into Map when scala type of map key is not String

### ISSUE THREAD

**Issue by rabik4** [gzoller/ScalaJack#12]
If key of the map is not String, following error is thrown:
java.lang.ClassCastException: java.lang.Integer 

Here is the test code: 
```
case class TestClass(myMap: Map[Int, Int])

it("Checking scalajack marshalling") {
  val myMap = Map(1 -> 1, 2 -> 3)
  val caseClass= TestClass(myMap)

  ScalaJack.render[TestClass](caseClass)
  val json = ScalaJack.render[TestClass](caseClass)
  val caseClassBack = ScalaJack.read[TestClass](json)
  
  caseClassBack.myMap.foreach { a =>
    try {
      println(s"a._1: ${a._1}")        // java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
    } catch {
      case t:Throwable => println(t.printStackTrace())
    }
  }
}
```

We were trying the marshall json into type of Map[Long, List[Long]] and encountered the bug.

**Comment by gzoller** [comment]
Hi Rabik!

Interesting idea but unfortunately not workable.  The effort in ScalaJack
would be immense.  Consider a hypothetical map:

Map[Map[List[Bar],Foo],Set[Map[Int,Blather]]

Argh!

Beyond the mess, though, the problem here isn't the ScalaJack
implementation, its the JSON spec.  See here: http://www.json.org/

According to the spec JSON object keys are strings and only strings.  There
may indeed be "street json" out there that doesn't follow the letter of the
spec, but technically it wouldn't be JSON anymore and you'd be in dangerous
waters if you tried to parse it with a non-ScalaJack parser.

This is also important because of ScalaJack's history (and Salat's, from
which ScalaJack was inspired).  There was a MongoDB integration as a major
feature, and with that I can't want to deviate from the JSON spec.  Have a
look at this API clip from MongoDBObject (Casbah):


   1.
   defiterator : Iterator[(String, AnyRef)]
   2. defkeySet : Set[String]
   3. defkeys : Iterable[String]


These are methods that navigate the keys.  Notice they're all strongly
typed as String.  So even if ScalaJack climbed the mountain to support
non-String keys, nothing in the pure JSON world would support it.

Greg

**Comment by gzoller** [comment]
No plan to address for now -- runs against JSON spec.

**Comment by gzoller** [comment]
Actually... I've just released ScalaJack 4.0 -- a total re-write.  Code is much simpler.  There are some changes to the api although the basic 2 legacy calls still work.  One new feature is that you can pass a VisitorContext object that has a param isCanonical.  Set this to false and you can have non-string keys in "json".  Of course you still can't save that to mongo, etc., but maybe for some uses it'll work.
